scripts/kfctl.sh

```diff
@@ -225,14 +225,13 @@ ksApply () {
     ks env add default --namespace "${K8S_NAMESPACE}"
   fi
 
-  # Create all the core components
-  ks apply default -c ambassador
-  ks apply default -c jupyterhub
-  ks apply default -c centraldashboard
-  ks apply default -c tf-job-operator
-  ks apply default -c argo
-  ks apply default -c katib
-  ks apply default -c spartakus
+  # Create the kubeflow application
+  # - first create all cluster resources
+  ks param set application bootstrap true
```

Comment:
i agree - there should be a check that all cluster resources were deployed. 

The concept is to separate the 'devops' part from the data scientist part.
That is - different data scientists could now deploy the application into 
different namespaces. Or different data scientists could share one namespace where 
the application is deployed by the 'owner'. We have a controller implementation that 
creates a namespace on behalf of a user. We've shown under RBAC rules that we use that 
each data scientists can not 'see' other data scientist's namespaces, jobs.
I can open this PR for review, though it is similar to #1769.

Reply:
> My suggestion would be to scope this PR to just providing an application resource whose sole purpose is to provide meta information about the Kubeflow deployment. This is high value because it will provide users visibility into Kubeflow.

Got it. So just generate the Application Kind based on what components have been generated, with an optional param if you want to generate an application for some of the components. `ks apply default -c application` would result in 
```yaml
all: [
  self.applicationCRD,
  self.application,
],
```
The application would hold information about all kf_app/ks_app/components unless specific components were provided eg `ks param set application components '["ambassador", "jupyterhub"]'`

Reply:
> So the infra team will want to deploy Kubeflow such that each user gets their own namespace. So within that namespace the user can launch their notebooks, models, jobs.

The user isolation mechanism we've prototyped has some overlap. 

If I understand correctly - we could have several global metacontrollers - notebook-controller, model-controller, job-controller. These metacontrollers are shared across teams, users. Each one would take a CRD from a user (perhaps emitted from a ui) and generate the child resources required into the namespace owned|shared by the user. For notebooks the notebook-controller would create child resources {service, deployment} where the container would be the notebook image. Would a separate application be instantiated per metacontroller and dependencies defined so there is an overall kubeflow-controller that is the ownerReference for everything?

Reply:
Correct. 

If you look at the controllers we have today

- JupyterHub
- Argo
- TFJob
- SeldonDeployment

These can be shared across users/teams. Each user/team can create their own resources Jupyter notebooks, Argo workflows, TFJobs, SeldonDeployments in their own namespace but share the controller.

I believe this has how most K8s built in controllers (e.g. ReplicaSet, Job controller work).

In some cases we have additional work to do to support multi-user. For example, in Jupyter I think we want to use ISTIO to restrict web access to each notebook so that only user A can access user A's notebook.